### DISCUSSION TITLE
Is there is any way to show ads on twitch?

### DISCUSSION THREAD

**Discussion by furrykit** [streamlink/streamlink#6129] [Q&A] [answered]
### Description

Is there is any way to show ads on twitch? I mean when ads starting, player just show "commercial break in progress" image, but not actual ads. When i open stream in VLC without streamlink, ads showing as it should, but when i try to launch vlc via streamlink its always "commercial break in progess" even if i use "--player-passthrough=hls" option and set useragent to default browser useragent
![photo_2024-08-12_10-19-45](https://github.com/user-attachments/assets/fba7168e-7e38-4fa7-843e-da95abec9cb1)

**Comment by bastimeyer** [comment, marked as answer]
> player just show "commercial break in progress" image
> Is there is any way to show ads on twitch?

This is a stupid idea, because this doesn't achieve anything and introduces forced stream discontinuities, which are unsupported, as you can read in the docs (first link).

- https://streamlink.github.io/cli/plugins/twitch.html
- https://streamlink.github.io/plugins.html#twitch
- https://streamlink.github.io/cli.html#twitch

Forcing ads requires setting special access token parameters, as well as obtaining a (new) client-integrity token:

- https://streamlink.github.io/latest/cli.html#cmdoption-twitch-access-token-param
- https://streamlink.github.io/latest/cli.html#cmdoption-twitch-force-client-integrity

[`--twitch-force-client-integrity`](https://streamlink.github.io/latest/cli.html#cmdoption-twitch-force-client-integrity) was added only recently and will be available in the upcoming 6.9.0 release, so until then, you'd need to use nightly builds or patch your twitch plugin version.

The following will cause real ads to be embedded into the stream, and if you want to watch ads every time, then also add [`--twitch-purge-client-integrity`](https://streamlink.github.io/latest/cli.html#cmdoption-twitch-purge-client-integrity).

```sh
$ streamlink \
  --twitch-access-token-param=playerType=site \
  --twitch-access-token-param=platform=web \
  --twitch-force-client-integrity \
  twitch.tv/CHANNEL best
```

**Reply by furrykit** [reply]
I download latest nighty build and trying to start streamlink like this --twitch-purge-client-integrity --twitch-access-token-param=playerType=site --twitch-access-token-param=platform=web --twitch-force-client-integrity "https://www.twitch.tv/mystreamurl" best --http-proxy=socks5h://127.0.0.1:61000
and got this errors

```
[cli][info] Found matching plugin twitch for URL https://www.twitch.tv/mystreamurl
[plugins.twitch][info] Removing cached client-integrity token...
[plugins.twitch][info] Acquiring new client-integrity token...
[webbrowser.webbrowser][info] Launching web browser: C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe (headless=False)
[plugins.twitch][error] Failed acquiring client integrity token
  + Exception Group Traceback (most recent call last):
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\plugins\twitch.py", line 620, in acquire
  |     client_integrity = CDPClient.launch(session, acquire_client_integrity_token)
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\cdp\client.py", line 170, in launch
  |     return trio.run(run_wrapper, strict_exception_groups=True)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "C:\Program Files\Streamlink\pkgs\trio\_core\_run.py", line 2306, in run
  |     raise runner.main_task_outcome.error
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\cdp\client.py", line 159, in run_wrapper
  |     async with cls.run(
  |   File "contextlib.py", line 210, in __aenter__
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\cdp\client.py", line 186, in run
  |     async with webbrowser.launch(headless=headless, timeout=timeout) as nursery:
  |   File "contextlib.py", line 231, in __aexit__
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\chromium.py", line 171, in launch
  |     async with super()._launch(self.executable, arguments, headless=headless, timeout=timeout) as nursery:
  |   File "contextlib.py", line 231, in __aexit__
  |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\webbrowser.py", line 87, in launch
  |     async with trio.open_nursery() as nursery:
  |   File "C:\Program Files\Streamlink\pkgs\trio\_core\_run.py", line 959, in __aexit__
  |     raise combined_error_from_nursery
  | ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\webbrowser.py", line 104, in launch
    |     yield nursery
    |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\chromium.py", line 172, in launch
    |     yield nursery
    |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\cdp\client.py", line 187, in run
    |     websocket_url = webbrowser.get_websocket_url(session)
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "C:\Program Files\Streamlink\pkgs\streamlink\webbrowser\chromium.py", line 182, in get_websocket_url
    |     return session.http.get(
    |            ^^^^^^^^^^^^^^^^^
    |   File "C:\Program Files\Streamlink\pkgs\requests\sessions.py", line 602, in get
    |     return self.request("GET", url, **kwargs)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "C:\Program Files\Streamlink\pkgs\streamlink\session\http.py", line 191, in request
    |     raise err from None  # TODO: fix this
    |     ^^^^^^^^^^^^^^^^^^^
    | streamlink.exceptions.PluginError: Unable to open URL: http://127.0.0.1:18227/json/version (SOCKSHTTPConnectionPool(host='127.0.0.1', port=18227): Max retries exceeded with url: /json/version (Caused by NewConnectionError('<urllib3.contrib.socks.SOCKSConnection object at 0x0000000004DCD490>: Failed to establish a new connection: 0x01: General SOCKS server failure')))
    +------------------------------------
[plugins.twitch][warning] No client-integrity token acquired
```

I think problem is only when using --http-proxy parameter, but i can't launch streams without proxy (because its blocked in my country). Any advice?

**Reply by bastimeyer** [reply]
It's a bug with the webbrowser API and HTTP session proxies. See #6130.
